File: src/main/kotlin/org/kite9/diagram/visualization/orthogonalization/flow/face/ConstraintGroup.java
package org.kite9.diagram.visualization.orthogonalization.flow.face;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import org.kite9.diagram.common.algorithms.det.DetHashSet;
import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.visualization.planarization.Face;

public class ConstraintGroup {

	private final Set<Edge> fixedConstraints;

	public ConstraintGroup(Set<Edge> fixedConstraints, int faces) {
		super();
		this.fixedConstraints = new DetHashSet<Edge>(fixedConstraints);
		this.faceCount = faces;
	}

	private Set<Edge> floatingConstraints = new DetHashSet<Edge>();

	private List<Route> requiredRoutes = new ArrayList<Route>();
	
	private int faceCount;

	public List<Route> getRequiredRoutes() {
		return requiredRoutes;
	}

	private Map<Face, List<Integer>> constrainedFaces = null;

	/**
	 * This is used in portion creation. It looks at which routes use this
	 * face, and asks for just the constraints within those routes on that
	 * face. This is to avoid breaking the face into too many unnecessary
	 * portions.
	 */
	public List<Integer> getConstraintsRequiredForFace(Face f) {
		if (constrainedFaces == null) {
			constrainedFaces = new HashMap<Face, List<Integer>>(faceCount * 2);
			for (Route r : requiredRoutes) {
				while (r!=null) {
					Face ff = r.face;
					List<Integer> constraintsForFace = constrainedFaces.get(ff);
					if (constraintsForFace == null) {
						constraintsForFace = new LinkedList<Integer>();
						constrainedFaces.put(ff, constraintsForFace);
					}
					
					if (!constraintsForFace.contains(r.in)) {
						constraintsForFace.add(r.in);
					}
					
					if (!constraintsForFace.contains(r.out)) {
						constraintsForFace.add(r.out);
					}
					
					Edge e1 = r.getInEdge();
					if (!fixedConstraints.contains(e1)) {
						floatingConstraints.add(e1);
					}
					
					Edge e2 = r.getOutEdge();
					if (!fixedConstraints.contains(e2)) {
						floatingConstraints.add(e2);
					}
					
					r = r.rest;
				}
			}
			
			for (Entry<Face, List<Integer>> ent : constrainedFaces.entrySet()) {
				Collections.sort(ent.getValue());
			}
		}
		
		return constrainedFaces.get(f);
	}

	public boolean isConstrained(Edge e) {
		return fixedConstraints.contains(e) || floatingConstraints.contains(e);
	}

	public String toString() {
		return fixedConstraints.toString();
	}
	
	public void addRoute(Route r) {
		requiredRoutes.add(r);
	}
}

File: visualization/kite9-visualization-common/src/jvmMain/kotlin/org/kite9/diagram/visualization/orthogonalization/flow/face/ConstraintGroup.kt
package org.kite9.diagram.visualization.orthogonalization.flow.face

import org.kite9.diagram.common.algorithms.det.DetHashSet
import org.kite9.diagram.common.elements.edge.Edge
import org.kite9.diagram.visualization.planarization.Face

class ConstraintGroup(private val fixedConstraints: Set<Edge>, private val faceCount: Int) {

    private val floatingConstraints: MutableSet<Edge> = DetHashSet()
    private val requiredRoutes: MutableList<Route> = ArrayList()

    fun getRequiredRoutes(): List<Route> {
        return requiredRoutes
    }

    private val constrainedFaces: Map<Face, List<Int>> by lazy {
        val out : MutableMap<Face, MutableList<Int>> = mutableMapOf()

        for (sr in requiredRoutes) {
            var rr : Route? = sr
            while (rr != null) {
                val ff = rr.face
                var constraintsForFace = out.get(ff)
                if (constraintsForFace == null) {
                    constraintsForFace = mutableListOf()
                    out[ff] = constraintsForFace
                }
                if (!constraintsForFace.contains(rr._in)) {
                    constraintsForFace.add(rr._in)
                }
                if (!constraintsForFace.contains(rr._out)) {
                    constraintsForFace.add(rr._out)
                }
                val e1 = rr.inEdge
                if (!fixedConstraints.contains(e1)) {
                    floatingConstraints.add(e1)
                }
                val e2 = rr.outEdge
                if (!fixedConstraints.contains(e2)) {
                    floatingConstraints.add(e2)
                }
                rr = rr.rest
            }
        }
        for ((_, value) in out) {
            value.sort()
        }

        out
    }

    /**
     * This is used in portion creation. It looks at which routes use this
     * face, and asks for just the constraints within those routes on that
     * face. This is to avoid breaking the face into too many unnecessary
     * portions.
     */
    fun getConstraintsRequiredForFace(f: Face): List<Int>? {
        return constrainedFaces[f]
    }

    fun isConstrained(e: Edge): Boolean {
        return fixedConstraints.contains(e) || floatingConstraints.contains(e)
    }

    override fun toString(): String {
        return fixedConstraints.toString()
    }

    fun addRoute(r: Route) {
        requiredRoutes.add(r)
    }
}